### ISSUE TITLE
Extend C# nullable analysis to check for nullable field/property assignments in methods called by constructors

### ISSUE THREAD

**Issue by Ramobo** [dotnet/roslyn#42531]
**Version Used**: C# 8.0, C# Tools 3.6.0-1.20127.6+572b1f0e0da95fd92af710ee58c95e2257cb08e3

**Steps to Reproduce**:
1. Enable nullable warnings;
2. Declare a non-nullable field or property;
3. Declare a constructor that calls a method;
4. In the method from step 3, initialize the field/property declared in step 2.

Example:
```CSharp
#nullable enable

public class MyClass
{
    public Exception myField;

    public MyClass() => Initialize(); // Warning CS8618: Non-nullable field 'myField' is uninitialized. Consider declaring the field as nullable.

    private void Initialize() => myField = new Exception();
}
```

**Expected Behavior**: The compiler detects that `myField` is initialized in `Initialize()`, which is called by the constructor, and does not issue a warning.

**Actual Behavior**: The compiler issues `Warning CS8618: Non-nullable field 'myField' is uninitialized. Consider declaring the field as nullable.` on the constructor's line.

**Comment by jcouv** [comment]
This was just [recently fixed](https://github.com/dotnet/roslyn/pull/41675) in 16.6p2 with an extension to the language for C# 9. 
Mark the `Initialize` method with `[System.Diagnostics.CodeAnalysis.MemberNotNull(nameof(myField))]` and set LangVersion to "preview", then the warning on the constructor will disappear.

**Comment by Ramobo** [comment]
An attribute for this just feels unclean. I don't know how the analysis works, but this seems like something too simple to require an attribute for, considering that it can analyze if/else branches.

**Comment by Ramobo** [comment]
@jcouv Seriously, can't this be automatically analyzed — without an attribute?

**Comment by jaredpar** [comment]
@Ramobo 

This type of analysis would require the compiler to do cross method analysis which is not done today. Doing so would be an enormous undertaking in the compiler. Further it would mean that the results of compilation differ when compiling against methods defined in source and those defined in assemblies (assemblies can be ref assemblies which means method bodies would be absent).

**Comment by RikkiGibson** [comment]
Related to #32358
